go-controller/pkg/clustermanager/clustermanager.go

```diff
@@ -0,0 +1,418 @@
+	defer func() {
+		if err != nil {
+			if errR := cm.clusterSubnetAllocator.ReleaseNodeSubnets(node.Name, allocatedSubnets...); errR != nil {
+				klog.Warningf("Error releasing node %s subnets: %v", node.Name, errR)
+			}
+		}
+	}()
+
+	// Set the HostSubnet annotation on the node object to signal
+	// to nodes that their logical infrastructure is set up and they can
+	// proceed with their initialization
+	return cm.updateNodeAnnotationWithRetry(node.Name, hostSubnets)
+}
+
+func (cm *ClusterManager) deleteNode(node *kapi.Node) error {
+	if config.HybridOverlay.Enabled {
+		cm.releaseHybridOverlayNodeSubnet(node.Name)
+	}
+
+	cm.clusterSubnetAllocator.ReleaseAllNodeSubnets(node.Name)
+	return nil
+}
+
+// hybridOverlayNodeEnsureSubnet allocates a subnet and sets the
+// hybrid overlay subnet annotation. It returns any newly allocated subnet
+// or an error. If an error occurs, the newly allocated subnet will be released.
+func (cm *ClusterManager) hybridOverlayNodeEnsureSubnet(node *kapi.Node, annotator kube.Annotator) (*net.IPNet, error) {
+	var existingSubnets []*net.IPNet
+	// Do not allocate a subnet if the node already has one
+	subnet, err := houtil.ParseHybridOverlayHostSubnet(node)
+	if err != nil {
+		// Log the error and try to allocate new subnets
+		klog.Infof("Failed to get node %s hybrid overlay subnet annotation: %v", node.Name, err)
+	} else if subnet != nil {
+		existingSubnets = []*net.IPNet{subnet}
+	}
+
+	// Allocate a new host subnet for this node
+	// FIXME: hybrid overlay is only IPv4 for now due to limitations on the Windows side
+	hostSubnets, allocatedSubnets, err := cm.hybridOverlaySubnetAllocator.AllocateNodeSubnets(node.Name, existingSubnets, true, false)
+	if err != nil {
+		return nil, fmt.Errorf("error allocating hybrid overlay HostSubnet for node %s: %v", node.Name, err)
+	}
+
+	if err := annotator.Set(hotypes.HybridOverlayNodeSubnet, hostSubnets[0].String()); err != nil {
+		_ = cm.hybridOverlaySubnetAllocator.ReleaseNodeSubnets(node.Name, allocatedSubnets...)
+		return nil, err
+	}
+
+	return hostSubnets[0], nil
+}
+
+func (cm *ClusterManager) releaseHybridOverlayNodeSubnet(nodeName string) {
+	cm.hybridOverlaySubnetAllocator.ReleaseAllNodeSubnets(nodeName)
+	klog.Infof("Deleted hybrid overlay HostSubnets for node %s", nodeName)
+}
+
+// shouldUpdateNode() determines if the ovn-kubernetes plugin should update the state of the node.
+// ovn-kube should not perform an update if it does not assign a hostsubnet, or if you want to change
+// whether or not ovn-kubernetes assigns a hostsubnet
+func shouldUpdateNode(node, oldNode *kapi.Node) (bool, error) {
+	newNoHostSubnet := util.NoHostSubnet(node)
+	oldNoHostSubnet := util.NoHostSubnet(node)
+
+	if oldNoHostSubnet && newNoHostSubnet {
+		return false, nil
+	} else if oldNoHostSubnet && !newNoHostSubnet {
+		return false, fmt.Errorf("error updating node %s, cannot remove assigned hostsubnet, please delete node and recreate", node.Name)
```

Comment:
I might very well be wrong, but shouldn't we swap this and the error that follows (line 414)?
(I also realise it's not your code, you took it from a different source file :)  )

Reply:
I'm not sure on this.

If you're sure, then I'll swap it. Please let me know :)

Reply:
@JacobTanenbaum I see the lines above were from a PR of yours from 2020. Do you agree we should swap the error messages? The way I read `if oldNoHostSubnet && !newNoHostSubnet` from above is: "previously the node had the "no host subnet" label, so we didn't allocate any host subnet to it, now the node lost this label and we should in theory allocate a host subnet to it (but we don't want to)", which matches the error message on line 414: `"error updating node %s, cannot assign a hostsubnet to already created node, please delete node and recreate"`. And viceversa. WDYT? Thanks! :)